api/src/main/java/io/grpc/ClientTransportFilter.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2023 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+/**
+ * Listens on the client transport life-cycle events. These filters do not have the capability
+ * to modify the channels or transport life-cycle event behavior, but they can be useful hooks
+ * for transport observability. Multiple filters may be registered to the client.
+ */
+@ExperimentalApi("https://gitub.com/grpc/grpc-java/issues/TODO")
+public abstract class ClientTransportFilter {
```

Comment:
"Filter." The server-side piece is called a Filter because it receives the transport's attributes and can change them. They are eventually viewable via ServerCall.getAttributes(). This allows you to add some data in transportReady and see it again in transportTerminated. That looks like it may be annoying to plumb here, but I'd have to stare at it a bit longer to tell.

I don't necessarily think we really need that though. We added the filtering for a specific purpose (caching data per-connection), which then went nowhere. Almost certainly it seems the filters would want to know the IP addresses and such of a transport.